### ISSUE TITLE
Logging Talon commands/state using ROS

### ISSUE THREAD

**Issue by kjaget** [FRC900/tasks#72]
The ROS control framework includes code to take state state from a controller and publish it as a node. This would be a great starting point for logging code - we could just run rosbag on the Jetson and dump the topic from this node to disk.

The code is in https://github.com/ros-controls/ros_controllers/blob/kinetic-devel/joint_state_controller/src/joint_state_controller.cpp.  First step is to clone the whole repo to a scratch dir and copy the joint_state_controller directory tree into our zebROS_ws/src directory.  

Probably makes sense to rename it to something like talon_state_controller.  Rename the directory.  Rename the code in the src/.cpp and rename include/joint_state_controller/joint_state_controller.h to include/talon_state_controller/talon_state_controller.h.  Then edit both those files to change the class name from JointStateController to TalonStateController and the namespace to talon_state_contoller.  Edit joint_state_plugin.xml to update the names in there to match.  Rename joint_state_plugin.xml to talon_state_plugin.xml.  Edit package.xml - change the owner fields to programmers@team900.org, reset the version to 0.0.1 or something, and so on.  Make sure to change the instance of joint_state_plugin.xml to talon_state_plugin.xml as well.  Edit CMakeLists.txt and change names from joint_state_* to talon_state_*.  

Then time to modify it a bit to handle Talon state.  Instead of using ROS's default JointStateInterface, we'd need to query TalonStateInterface. The definition of this is in zebROS_ws/src/talon_interface/include/talon_interface/talon_state_interface.h. Specifically, the data fields tracked are in the TalonHWState class. Right now the class is kinda sparse but as we add more state it will be included there.

We'd also have to create our own message type, with entries matching the data in TalonHWState.  Then, in the loop in update(), populate each field using a call to the correct get() method in TalonHWState and publish.

TBD - do we want to publish things which will never or rarely change (e.g. can_id, limit switch config, etc)?

**Comment by kjaget** [comment]
Another plus in getting this stuff published as a topic is that we can view it visually using http://wiki.ros.org/rqt_plot on any machine attached to the network...

**Comment by FalconWarriorr** [comment]
Yes I think that is definitely a useful functionality, but that we also need the capability of creating a modular dashboard capable of viewing multiple different types of data at any time and also data from any previous match

**Comment by kjaget** [comment]
Remember that rosbag can replay saved data.  And nodes can read the replayed data as if it were live

Simplest way to get there is to write nodes which read replay data (or live data, same thing) and convert on the fly to whatever output format is needed for the plotting software you all come up with.

**Comment by kjaget** [comment]
@FalconWarriorr - can you add 2017Preseason/zebROS_ws/src/talon_state_controller/CMakeLists.txt to git and push it.  You might have to do "add -f" to force it to be added.

Thanks!

**Comment by kjaget** [comment]
OK, to test this:

There's a file which contains the config for the all the controller stuff : https://github.com/FRC900/ros_control_boilerplate/blob/kinetic-devel/config/robot_in_a_box.yaml

The file has entries for controllers to load. Right now it loads joint_state_controller, but that should be replaced by talon_state_controller.  Look at line 42.  This defines a controller named joint_state_controller, and then creates it as a type joint_state_controller/JointStateController.  The type should be changed to talon_state_controller/TalonStateController.  The name could stay the same (it is arbitrary) but it makes sense to make the name be what the controller actually is, so change that also.

The type is defined in talon_state_controller/talon_state_plugin.xml.  That files looks like it is almost there.  The name in there matches the type listed above, which is good. You'll have to change the path (1st line) to be lib/libtalon_state_controller.  This will match the output file name for the code, which is defined in CMakeLists.txt.  I believe CMakeLists is already correct - the add_library call defines the output name, and that is based on PROJECT_NAME which is set as talon_state_controller already.

Re-build using catkin_make after making those changes.  Try to run it using

roslaunch ros_control_boilerplate sim_in_a_box.launch

If you were running on real hardware, you'd use robot_in_a_box.launch. Since we have a limited amount of that, we made up a simulated one. It uses the same robot_in_a_box.yaml config but runs code which doesn't try to write to non-existent robot hardware.

A rostopic list should show the talon state topic being published.  And a rostopic echo should show they data in the correct format.

**Comment by kjaget** [comment]
More data has been added to TalonHWState - update the talon_state_controller to match : 

			double output_current_;
			double bus_voltage_;
			double pidf_p_[2];
			double pidf_i_[2];
			double pidf_d_[2];
			double pidf_f_[2];
			unsigned pidf_izone_[2];
			int closed_loop_error_; //this is an int :)
			int fwd_limit_switch_closed_;
			int rev_limit_switch_closed_;
			TalonMode talon_mode_;
                        int slot_;
